dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/sam/SamPermissionsImpl.java

```diff
@@ -171,6 +178,26 @@ public void removePermission(User user, Workflow workflow, String email, Role ro
         }
     }
 
+    /**
+     * Derives <code>user</code>'s {@link Permission} (email + role) on <code>workflow</code>, without
+     * being able to read the SAM permissions directly, by calling the SAM
+     * APIs to see what actions the user can perform on the workflow.
+     *
+     * @param user
+     * @param workflow
+     * @return
+     */
+    private Optional<Permission> derivePermission(User user, Workflow workflow) {
+        // If we gotten this far, we can safely assume there is a Google TokenType
```

Comment:
could do an assert, it wouldn't be enforced, but I find it more clear for essentially, preconditions

Reply:
👍